### DISCUSSION TITLE
FIR Filter with coefficients count > 137

### DISCUSSION THREAD

**Discussion by Sobhan-Nami** [pschatzmann/arduino-audio-tools#966] [Q&A] [answered]
Hi again.
I'm sorry to ask a lot of questions.

I use FIR Filter with coefficients created by the https://fiiir.com/ site that you suggested.
when I use a filter with a coefficient count of more than 200 the filter work weirdly.

I use one source file for all of the filters and just replace Coef in the code and use Sonic-Visualizer to check the spectrum.

this is the code I used for filtering.
reading wav file from sd -> decode -> filter -> encode -> write back
<br/>
```
File src;
File des;
AudioInfo info;
EncodedAudioStream wavEncoder(&des, new WAVEncoder());
FilteredStream<int16_t, float> inFiltered(wavEncoder, 1);
EncodedAudioStream wavDecoder(&inFiltered, new WAVDecoder());
StreamCopy fileCopier(wavDecoder, src);
#define FILTER_SIZE  185

const float coef[FILTER_SIZE] = {};

void convertData()
{

    File root = SD_MMC.open(SOUND_DIR);
    info.channels = 1;
    info.sample_rate = 44100;

    String fileName = root.getNextFileName();
    wavEncoder.begin(info);
    inFiltered.setFilter(0, new FIR<float>(coef));
    wavDecoder.begin(info);

    int index = fileName.lastIndexOf('/');
    int length = fileName.length();
    String uploadedFilename = UPLOADED_DIR + fileName.substring(index, length);

    Serial.println("filename: " + fileName + " start filtering");
    Serial.println("new file in : " + uploadedFilename);

    src = SD_MMC.open(fileName);
    des = SD_MMC.open(uploadedFilename, FILE_WRITE);
    fileCopier.copyAll();
    Serial.println("All Done");

    src.close();
    des.close();

}

void setup()
{
    Serial.begin(115200);
    mountSD();

    AudioLogger::instance().begin(Serial, AudioLogger::Warning);
    Serial.println(WiFi.localIP());
}
bool first = true;
void loop()
{
    if (first)
    {
        convertData();
        first = false;
    }

}
```
<br/>

### This is a low-pass filter with a 2k cutoff and 137 coef (coef used in [this example](https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/tests/test-filter/test-filter.ino))

![low-pass with 2k-137](https://github.com/pschatzmann/arduino-audio-tools/assets/24646960/ec8d8b22-b6f7-49f1-bbbb-521eafcbea01)
<br/>

### This is a high-pass filter with a 3k cutoff and 137 coef. 

![high-pass with 3k-137](https://github.com/pschatzmann/arduino-audio-tools/assets/24646960/0fd83024-c221-461d-a07d-08bf3e5320b1)
<br/>

### This is a band-pass filter with a 3k-6k cutoff and 185 coef.
![band-pass 3k-6k 185](https://github.com/pschatzmann/arduino-audio-tools/assets/24646960/d0ece2e7-1683-4071-8882-197112add216)

<br/>

### and the last one is a band-pass filter with a 3k-6k cutoff and 621 coef.
![band-pass 3k-6k 620](https://github.com/pschatzmann/arduino-audio-tools/assets/24646960/59ae3ce4-b790-455d-8fba-6972cc2ee189)

**Comment by pschatzmann** [comment, marked as answer]
Microcontrollers are very slow and the challenge is always to provide functionality within limited resources.  
You need to make sure that you can provide the result within your requested sampling rate.

Solutions
- Keep the number of coefficients as small as possible.
- Or move the processing to the desktop which has much faster floating point operations.
- Decrease the sample rate

I assume you were using an ESP32, so if you use other microcontrollers the limit will be much lower.

**Reply by Sobhan-Nami** [reply]
Ok.
Thank you for your reply.
